snuba/query/processors/custom_function.py

```diff
@@ -0,0 +1,74 @@
+from dataclasses import replace
+from typing import Sequence
+
+from snuba.query.exceptions import InvalidExpressionException
+from snuba.query.expressions import Column, Expression, FunctionCall
+from snuba.query.logical import Query
+from snuba.query.parser.expressions import parse_clickhouse_function
+from snuba.query.processors import QueryProcessor
+from snuba.request.request_settings import RequestSettings
+
+
+class InvalidCustomFunctionCall(InvalidExpressionException):
+    def __str__(self) -> str:
+        return f"Invalid custom function call {self.expression}: {self.message}"
+
+
+class CustomFunction(QueryProcessor):
+    """
+    Defines a custom snuba function.
+    The custom function has a name, a signature in the form of a list of
+    parameter names and an expanded expression which is the body of the
+    custom function.
+
+    The custom function is invoked in a query as a standard snuba function.
+
+    Example:
+    CustomFunction(name="power_two", param_names=[value], body="value * value")
+
+    would transform
+    `["power_two", ["f", ["something"]], "alias"]`
+    into
+    `f(something) * f(something) AS alias`
+
+    Would raise InvalidCustomFunctionCall if a custom function is invoked with
+    the wrong number of parameters.
+    """
+
+    def __init__(self, name: str, param_names: Sequence[str], body: str) -> None:
```

Comment:
What do you think about including the param types here as well and validating them?

Reply:
I am not sure about that, It would only work for bare columns, while a user can pass a complex function in we cannot really infer the type for.
The validators I built before are best effort in order to catch common errors, I think that adding them into customizations may be confusing since they would be best effort there as well. I can try. What do you think ?

Reply:
I'm not sure exactly what you meant by inferring the type. I was thinking of something pretty simple, like for example you would specify that the Apdex function should take 3 arguments - a string (or column),  int and int and just doing a quick check on that.

Reply:
For inferring the type I mean that, if the parameter passed by the user is a raw column, or a literal then I can say whether that column is of the right type:
```
["apdex", ["'invalid'", 100, 200]]
```
`invalid` is not a column, so we can reject.

Though, if the user passes a function as parameter:
```
["apdex", [["assertNotNull", ["column"]], 100, 200]]
```
We do not know what the type of "assertNotNull" is without building a type inference system. Which means building a type checker.

Reply:
Oh right, makes sense. Would be nice to have at some point for sure.

Reply:
validation added